docs/source/cpp/acero/developer_guide.rst

```diff
@@ -0,0 +1,692 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+.. cpp:namespace:: arrow::acero
+
+=================
+Developer's Guide
+=================
+
+This page goes into more detail into the design of Acero.  It discusses how
+to create custom exec nodes and describes some of the philosophies behind Acero's
+design and implementation.  Finally, it gives an overview of how to extend Acero
+with new behaviors and how this new behavior can be upstreamed into the core Arrow
+repository.
+
+Understanding ExecNode
+======================
+
+ExecNode is an abstract class with several pure virtual methods that control how the node operates:
+
+:func:`ExecNode::StartProducing`
+--------------------------------
+
+This method is called once at the start of the plan.  Most nodes ignore this method (any
+neccesary initialization should happen in the construtor or Init).  However, source nodes
+will typically provide a custom implementation.  Source nodes should schedule whatever tasks
+are needed to start reading and providing the data.  Source nodes are usually the primary
+creator of tasks in a plan.
+
+.. note::
+   The ExecPlan operates on a push-based model.  Sources are often pull-based.  For example,
+   your source may be an iterator.  The source node will typically then schedule tasks to pull one
+   item from the source and push it into the plan.
```

Comment:
```
   item from the source and push it into the plan.
```

Should here be `into the node`?

Reply:
Yes.  This is a little imprecise.  I've updated to:

```
   The ExecPlan operates on a push-based model.  Sources are often pull-based.  For example,
   your source may be an iterator.  The source node will typically then schedule tasks to pull one
   item from the source and push that item into the source's output node (via ``InputReceived``).
   ```